[Free trial] Implement and show Free Trial banner and badges in the App after Free Trial starts

pac-guerreiro opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/43672
PROPOSAL: N/A


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. On web, write this on the browser’s developer console: 
```
window.Onyx.multiSet({
       [ONYXKEYS.NVP_FIRST_DAY_FREE_TRIAL]: <TODAY>,
       [ONYXKEYS.NVP_LAST_DAY_FREE_TRIAL]: <6 DAYS FROM NOW>,
})

// Replace <TODAY> and <6 DAYS FROM NOW> using 'yyyy-MM-dd HH:mm:ss' format
```
2. On the homepage, verify that the trial duration badge is shown in front of Expensify chat
3. Select the Expensify chat and verify that the trial duration badge is shown at the right of the chat header
4. Then, go to the settings page and verify that the trial duration badge is shown in front of the subscriptions option, under workspace section
5. Then, go to `/settings/subscription` and verify that the trial duration banner is shown in the card section
6. Don’t forget to write this on the browser’s developer console in order to clear the free trial UI: 
```
window.Onyx.multiSet({
       [ONYXKEYS.NVP_FIRST_DAY_FREE_TRIAL]: null,
       [ONYXKEYS.NVP_LAST_DAY_FREE_TRIAL]: null,
});
```

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

Same as above

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
7. Verify a modal appears displaying a preview of that image
--->

Same as above

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->
![Android - Native - Chat list](https://github.com/Expensify/App/assets/48553277/e2cc8169-da37-4cc1-b719-42ac7736e9a0)
![Android - Native - Expensify DM](https://github.com/Expensify/App/assets/48553277/485cb8ad-e06a-40f3-98ad-f848589aec6c)
![Android - Native - Settings](https://github.com/Expensify/App/assets/48553277/5be44585-afb8-4c9e-83c7-35c2911d78d0)
![Android - Native - Subscription](https://github.com/Expensify/App/assets/48553277/dede0a70-3113-4961-b256-cec244f82a0d)

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->
![Android - Chrome - Chat list](https://github.com/Expensify/App/assets/48553277/9b51efc7-2418-412c-851c-33b06f7dc070)
![Android - Chrome - Expensify DM](https://github.com/Expensify/App/assets/48553277/8cbe205f-1236-4fdd-b77f-778cd6fa2d2e)
![Android - Chrome - Settings](https://github.com/Expensify/App/assets/48553277/b434c6e6-8c2e-47a0-b0bf-854ac952ee1d)
![Android - Chrome - Subscription](https://github.com/Expensify/App/assets/48553277/e0d13cd5-917f-4e12-9507-47db724d12ed)

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->
![iOS - Native - Chat list](https://github.com/Expensify/App/assets/48553277/56d059f7-48d6-405a-b32b-6ab4b5c31a06)
![iOS - Native - Expensify DM](https://github.com/Expensify/App/assets/48553277/a2a296e3-95a5-4bdd-8aaa-9f6193d7f9a8)
![iOS - Native - Settings](https://github.com/Expensify/App/assets/48553277/a43ab50d-35d9-476e-886a-07863b837ba2)
![iOS - Native - Subscription](https://github.com/Expensify/App/assets/48553277/d1619785-fcd0-4aaa-b088-1e2e11c44c4f)

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->
![iOS - Safari - Chat list](https://github.com/Expensify/App/assets/48553277/b3c66886-d7e7-49c1-a871-68987a1ebd9c)
![iOS - Safari - Expensify DM](https://github.com/Expensify/App/assets/48553277/56ab4407-7f89-4532-9106-a11d05db81cc)
![iOS - Safari - Settings](https://github.com/Expensify/App/assets/48553277/e268b4ae-901e-4950-ab3d-6a1884dec3b5)
![iOS - Safari - Subscription](https://github.com/Expensify/App/assets/48553277/27eaa6be-d54f-4e0c-aec9-2b9f4640e94d)

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->
![MacOS - Chrome - Expensify DM](https://github.com/Expensify/App/assets/48553277/de88d712-53d1-436c-a273-db5ab6ac912a)
![MacOS - Chrome - Subscription](https://github.com/Expensify/App/assets/48553277/3c62270c-14c6-4b11-ae8f-f191bbd065e6)

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->
![MacOS - Native - Expensify DM](https://github.com/Expensify/App/assets/48553277/9a9e14ab-bd09-44ed-84f9-6c99284ce82b)
![MacOS - Native - Subscription](https://github.com/Expensify/App/assets/48553277/4f0bcb67-dcb8-4bc2-9b32-49910f86d6cc)

</details>

pac-guerreiro committed:
feat: implement free trial banner and badges

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/43672-implement-and-show-free-trial-banner-and-badges-after-free-trial-starts

# Conflicts:
#	src/pages/settings/Subscription/CardSection/CardSection.tsx

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/43672-implement-and-show-free-trial-banner-and-badges-after-free-trial-starts

rezkiy37 commented on src/styles/utils/sizing.ts:
```diff
@@ -34,6 +34,10 @@ export default {
         maxHeight: '100%',
     },
 
+    mnh0: {
+        minHeight: 0,
+    },
+
```
This style already exists on line 49.

pac-guerreiro committed:
refactor: remove duplicate sizing

fabioh8010 commented on src/pages/settings/Subscription/CardSection/BillingBanner/TrialBillingBanner.tsx:
Could we rename to `TrialStartedBillingBanner`?

fabioh8010 commented on src/pages/settings/Subscription/CardSection/BillingBanner/TrialBillingBanner.tsx:
```diff
@@ -0,0 +1,22 @@
+    );
+}
+
+TrialBillingBanner.displayName = 'PreTrialBillingBanner';
```
```suggestion
TrialStartedBillingBanner.displayName = 'TrialStartedBillingBanner';
```

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/43672-implement-and-show-free-trial-banner-and-badges-after-free-trial-starts

pac-guerreiro committed:
refactor: apply suggestions

pac-guerreiro committed:
refactor: apply suggestion from https://github.com/Expensify/App/pull/43855#discussion_r1651583350

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/43672-implement-and-show-free-trial-banner-and-badges-after-free-trial-starts

# Conflicts:
#	src/pages/settings/Subscription/CardSection/CardSection.tsx

pac-guerreiro committed:
refactor: apply suggestions

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/43672-implement-and-show-free-trial-banner-and-badges-after-free-trial-starts

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/43672-implement-and-show-free-trial-banner-and-badges-after-free-trial-starts

pac-guerreiro committed:
fix: subscription badge border color

pac-guerreiro commented:
> Bug 1: The border of the free Trial badge in Subscription's menu item is not as bold as the design doc
> 
> ![Screenshot 2024-06-27 at 04 06 05](https://private-user-images.githubusercontent.com/9639873/343334251-5154fe03-bee6-4c12-8339-77c3601a2161.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk1MDkwMDcsIm5iZiI6MTcxOTUwODcwNywicGF0aCI6Ii85NjM5ODczLzM0MzMzNDI1MS01MTU0ZmUwMy1iZWU2LTRjMTItODMzOS03N2MzNjAxYTIxNjEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjdUMTcxODI3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTA4NmVkYjg0NTgxYTg0N2ViZDNmOGQ1ZTE1NDU0ODQyZjJiMTEyZTc1ODkxZjE1OTFmMjhkZDgxMWI4OWFiMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.LSWqKv8dhKttoBqw8mx5J2PY5MTjiP4qhe_V6Sr9lJY)
> 
> Moreover, the text "Subscription" is cut off on Web

@hoangzinh I fixed the border color issue! But the issue with the text getting cropped is expected, because this is caused by the right icon that is shown currently, since the subscription button redirects the user to the old dot subscription page.
When the developer that’s going to change this behaviour so that the user sees the new dot subscription page, this issue will disappear!I fixed the border color issue! But the issue with the text getting cropped is expected, because this is caused by the right icon that is shown currently, since the subscription button redirects the user to the old dot subscription page.
When the developer that’s going to change this behaviour so that the user sees the new dot subscription page, this issue will disappear!

trjExpensify commented:
@dannymcclain can you give this a looksie today as well, as you lead the design on it? Thanks!

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/43672-implement-and-show-free-trial-banner-and-badges-after-free-trial-starts

pac-guerreiro committed:
fix: applying badge success style out of free trial period

pac-guerreiro commented on src/pages/settings/InitialSettingsPage.tsx:
```diff
@@ -209,6 +212,8 @@ function InitialSettingsPage({session, userWallet, bankAccountList, fundList, wa
                 iconRight: Expensicons.NewWindow,
                 link: () => Link.buildOldDotURL(CONST.OLDDOT_URLS.ADMIN_POLICIES_URL),
+                badgeText: SubscriptionUtils.isUserOnFreeTrial() ? translate('subscription.badge.freeTrial', {numOfDays: SubscriptionUtils.calculateRemainingFreeTrialDays()}) : undefined,
+                badgeStyle: styles.badgeSuccess,
```
@hoangzinh good point! Suggestion applied 😄

pac-guerreiro commented:
@hoangzinh feedback applied 😄

dannymcclain commented:
This is looking good! Small suggestions: can we remove the word "below" from the message about adding a payment card? That way it will make more sense when viewing in native apps (since there's no option to add a payment card).

### Current
**Free trial: X days left!**
Add a payment card below to continue using all of your favorite features.

### Proposal
**Free trial: X days left!**
Add a payment card to continue using all of your favorite features.

The banner is so close to the big add payment card button that I don't really think we need to say "below" in that message.

chiragsalian reviewed (dismissed):
Code LGTM, not merging since danny has a change requested [here](https://github.com/Expensify/App/pull/44371#issuecomment-2197152210).
Let me know once addressed.

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/43672-implement-and-show-free-trial-banner-and-badges-after-free-trial-starts

pac-guerreiro committed:
feat: change free trial started banner copy

pac-guerreiro commented:
@dannymcclain I just pushed the change you requested 😄 

@chiragsalian I think we're good to go now!

pac-guerreiro commented:
@hoangzinh the spanish translation doesn't translate the word `below`

chiragsalian commented:
Looks like you've got some merge conflicts. Can you address them.

pac-guerreiro committed:
Merge branch 'main' into pac-guerreiro/feature/43672-implement-and-show-free-trial-banner-and-badges-after-free-trial-starts

# Conflicts:
#	src/pages/settings/Subscription/CardSection/CardSection.tsx

pac-guerreiro commented:
@chiragsalian conflicts resolved 😄

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/9783702773) to staging by https://github.com/chiragsalian in version: 9.0.4-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Cherry-picked](https://github.com/Expensify/App/actions/runs/9800117961) to staging by https://github.com/tgolen in version: 9.0.4-5 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

@Expensify/applauseleads please QA this PR and check it off on the [deploy checklist](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3AStagingDeployCash) if it passes.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/9872741083) to production by https://github.com/Julesssss in version: 9.0.5-13 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/9945901401) to production by https://github.com/thienlnam in version: 9.0.6-8 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅
